engine/engine_test.go

```diff
@@ -347,6 +347,242 @@ func TestEngine_NotifyPutThenNotifyRemove(t *testing.T) {
 	require.NotEqual(t, gotLatestAfterRmAdCid, gotLatestAdCid)
 }
 
+func TestEngine_WhenPublishForMultipleProviders_AdsShouldBeOnTheSameChain(t *testing.T) {
+	ctx := contextWithTimeout(t)
+	rng := rand.New(rand.NewSource(1413))
+
+	mhs1 := testutil.RandomMultihashes(t, rng, 42)
+	mhs2 := testutil.RandomMultihashes(t, rng, 42)
+
+	provider1id := "Provider 1"
+	provider1Addrs := []string{"/ip4/0.0.0.0/tcp/1234/http"}
+	provider2id := "Provider 2"
+	provider2Addrs := []string{"/ip4/0.0.0.0/tcp/4321/http"}
+
+	subject, err := engine.New()
+	require.NoError(t, err)
+	err = subject.Start(ctx)
+	require.NoError(t, err)
+	defer subject.Shutdown()
+
+	wantContextID1 := []byte("fish")
+	wantContextID2 := []byte("bird")
+	subject.RegisterMultihashLister(func(ctx context.Context, p string, contextID []byte) (provider.MultihashIterator, error) {
+
+		if string(contextID) == string(wantContextID1) {
+			return provider.SliceMultihashIterator(mhs1), nil
+		} else if string(contextID) == string(wantContextID2) {
+			return provider.SliceMultihashIterator(mhs2), nil
+		}
```

Comment:
Would it make sense to make the conditions stronger by also asserting the passed provider ID  is the expected one for the given context ID?

Reply:
good point, added to the condition